### ISSUE TITLE
How do we mock useQuery

### ISSUE THREAD

**Issue by skywickenden** [nearform/graphql-hooks#400]
### Description

It would be great if there was a short doc on how to easily mock useQuery when writing tests for pages that use it.

### Suggested implementation

At present I mock the graphql-hooks import and then conditionally return the correct query, but this is messy. Is there a better way?

    jest.mock('graphql-hooks', () => {
      return {
        useQuery: (query, options) => {
          // ... some logic to decide which query has been requested.
          return {
            loading: false,
            error: null,
            data: require('./some_mock_query_data').default,
          }
        },
      }
    })

**Comment by lynxtaa** [comment]
For our project we just mocked fetch function in GraphQLProvider. Hope it helps

```jsx
// ./components/__mocks__/GraphQLProvider.js
import { useMemo } from 'react'
import { GraphQLClient, ClientContext } from 'graphql-hooks'

export const getResult = jest.fn()

function GraphQLProvider({ children }) {
  const client = useMemo(
    () =>
      new GraphQLClient({
        url: '/',
        onError: jest.fn(),
        fetch: (url, { body }) =>
          Promise.resolve({
            ok: true,
            json: () => getResult(_.isString(body) ? JSON.parse(body) : body),
          }),
      }),
    [],
  )

  return <ClientContext.Provider value={client}>{children}</ClientContext.Provider>
}

export default GraphQLProvider

// setupTests.js
jest.mock('./components/GraphQLProvider')

// User.js
import React from 'react'
import { useQuery } from 'graphql-hooks'

const GET_USER_BY_ID = `
  query getUserById($userId: Int!) {
    user(id: $userId) {
      id
      firstname
    }
  }
`

function User({ id }) {
  const { data } = useQuery(GET_USER_BY_ID, { variables: { userId: id } })

  return <div>{data ? data.user.firstname : 'Loading...'}</div>
}

// User.test.js
import { render, waitForElement } from '@testing-library/react'

import { getResult, GraphQLProvider } from './components/GraphQLProvider'
import User from './User'

const withProviders = ({ children }) => <GraphQLProvider>{children}</GraphQLProvider>

it('should render user info', async () => {
  getResult.mockResolvedValue({ data: { user: { id: 1, firstname: 'John' } } })

  const { getByText } = render(<User id={1} />, { wrapper: withProviders })

  const trigger = await waitForElement(() => getByText('John'))

  expect(getResult).toHaveBeenCalledWith({
    query: expect.stringContaining('getUserById'),
    variables: { userId: 1 },
  })
})
```

**Comment by akashkrishnan98** [comment]
How do I mock data for child components while testing the parent component?

Eg:
```
<User> // mock getUserById
  <UserAvatar /> // How to mock getUserAvatar while testing <User />
</User>
```

**Comment by simoneb** [comment]
I would mock UserAvatar. If you're testing User in a unit test, it's a good idea to mock what its dependencies do. If you want to test User and UserAvatar together, then you'll need to mock useQuery.

**Comment by akashkrishnan98** [comment]
Let me rephrase my question:

I have multiple graphql queries for <User />.
How do I execute these? How can I specifically mock each query within the same component?

**Comment by simoneb** [comment]
You rely on your mocking library to change its behavior depending on the query that's being executed.

**Comment by chrisfrancis27** [comment]
In case anyone else stumbles across this, here's a very basic example of return different mock data depending on the query:
```js
// setupTests.ts
jest.mock("graphql-hooks", () => {
  return {
    useManualQuery: (query: string) => {
      let data = {};

      if (query.match("query blogPosts")) {
        data = require("../test/mocks/blogposts.json");
      } else if (query.match("query commentsByBlogPost")) {
        data = require("../test/mocks/comments.json");
      } else if (query.match("query countries")) {
        data = require("../test/mocks/countries.json");
      }

      return [
        () =>
          Promise.resolve({
            loading: false,
            data,
          }),
        {
          loading: false,
          data,
        },
      ];
    },
  };
});
```
